Fixes #32035 - Add DSL autocompletion in templates

ofedoren opened the pull request:
This PR introduces autocompletion for templates writing. It's far for being perfect, but still seems to be useful already.

Current additions:
 - Basic autocompletion. Supports built-in Ruby/Rails `keywords`, local variables.
 - Parses `/templates_doc` to get macros/properties as snippets.
 - Shows related information about macro, -- signature, short description, examples, return value, -- as a tooltip.
 - Snippets also have default values set as per documentation
 
 Current issues:
  - Sometimes the related info is not being shown fully; probably the issue is in `brace` that doesn't consider user's `viewport` for `docTooltip`.
  - This is more editor with autocompletion than IDE, so there is no `object.suggestions` pattern that would show what methods can be called on what objects, i.e. you cannot type `@host.` and get only the host methods. Current workaround is that properties are shown with prefixes, so one need to type `prefix#property` to get the snippet, e.g. `Managed#name`.
  - Doesn't check if passed value type is valid .
  - `ApipieDSL.docs` is used as a documentation to get the macros from. This means: the latest changes in code/docs, english (or default lang) only, default version. We can also use stored cache or any combinations.
  
  RFCs:
   - Implement IDE-like behaviour based on return values. Note: probably will require lots of js code on client side.
   - Enhance visuals for autocompletion.
   - Restrict snippets and show only related ones for each kind of template, i.e. in `report template` editor show only macros from `Reports` section. This can obviously reduce unrelated macros, but this needs review of all sections to ensure everything is properly scoped; probably will require some restructures in the documentation. 
   
All reviews, thoughts and suggestions are more than welcome to push the progress with it.

ares commented:
Wow, this is really awesome. I know it's not perfect but hopefully can be tweaked a little. The main thing that is quite confusing is, is the list of keywords. These are typically things user won't use such as stdin, assert_match and many other things that are disallowed in safemode. I think that limiting the autocompletion to just properties and macros would be a big win.

Another thing that I immediatelly experienced is typing outside of erb tag. E.g. when I edit kickstart file, adding more packages to %packages section is annoying (one work per line, enter selected to autocompletion so I always have to press scape).

Btw you can mark the PR as draft when you open it. Also a good next step would be to start the RFC conversation on discourse, I'll move the above comments there if you open it.

lzap commented:
This is totally demo-worthy, nice. We need UI experts here.

ofedoren commented:
@ares, thanks! You might want to share more thoughts here: https://community.theforeman.org/t/rfc-autocompletion-in-templates/23155

ofedoren committed:
Fixes #32035 - Add DSL autocompletion in templates

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/EditorHelpers.js:
```diff
@@ -0,0 +1,162 @@
+const resolveDefault = param => {
+  const value = param.default;
+
+  switch (typeof value) {
```
```suggestion
const resolveDefault = ({ default }) => {
  switch (typeof default) {

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/EditorHelpers.js:
```diff
@@ -0,0 +1,162 @@
+};
+
+const methodReturns = method => {
+  const retObject = method.returns.object;
```
Can you destructurt `method` or `retObject`?

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorSettings.js:
```diff
@@ -74,6 +76,24 @@ const EditorSettings = ({
         </div>
+        <div className="cog-popover-dropdown">
+          <div className="cog-popover-dropdown-title">
+            {__('Autocompletion')}
```
can you add it here?
https://github.com/theforeman/foreman/blob/2bafa9415a157d683c4352c9aaf097a1119eedd8/.eslintrc#L13

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/EditorConstants.js:
```diff
@@ -34,3 +34,4 @@ export const EDITOR_MODES = [
   'Xml',
   'Yaml',
 ];
+export const EDITOR_AUTOCOMPLETIONS = ['On', 'Off'];
```
These should be translated. not sure if here or in a different place. Also when these are translated this line should change to fit:
`enableBasicAutocompletion={autocompletion === 'On'}`

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorSettings.js:
```diff
@@ -74,6 +76,24 @@ const EditorSettings = ({
+          <div className="cog-popover-dropdown-title">
+            {__('Autocompletion')}
+          </div>
+          <Dropdown id="autocompletion-dropdown">
```
Maybe instead on/off it can be a checkbox? :)

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/EditorHelpers.js:
```diff
@@ -0,0 +1,162 @@
+  if (!cache) return;
+
+  const langTools = ace.acequire('ace/ext/language_tools');
+  // Custom basic snippets
```
How will a user know about these?

ofedoren replied:
I guess there is no way better then to try to write `tg` to see ERB related tags/macros. I thought about a special character, e.g. `:` or `!`, but not sure if it's a good idea...

MariaAga replied:
If it's common practice in templates then no changes are needed, maybe just mention it in the release notes :)

MariaAga replied:
Didn't notice until now that there is a "help" tab in the templates, maybe add it there?

ofedoren replied:
Regarding this: after few changes and the fact that the autocompletion is not "live" by default, a user will see all the available tags at the top of the list when presses `Ctrl + Space`:
![ScreenShot-1633699735464](https://user-images.githubusercontent.com/32508194/136565752-62ce7f06-4ad4-4736-b522-fca329cf68f4.png)

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/EditorHelpers.js:
```diff
@@ -0,0 +1,162 @@
+  ];
+  /* eslint-enable */
+  // Parse JSON with DSL documentation
+  forEach(JSON.parse(cache).docs.classes, (cls, clsId) => {
```
clsId isn't used, can it be removed?

MariaAga reviewed (changes requested):
Awesome pr, left some comments
Also I think adding:
```
.ace_tooltip .ace_doc-tooltip{
position: absolute;
}
```
Will fix the issue where the tooltip is hidden when its larger than the screen

ofedoren commented:
Thanks, @MariaAga :rose:

I've updated PR based on resolved conversations. Not sure what to do with the last one though...

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/editor.scss:
```diff
@@ -314,3 +314,14 @@
+}
+
+.ace_tooltip,
+.ace_doc-tooltip {
```
Should be `.ace_tooltip.ace_doc-tooltip` As the tooltip has both classes, I accidentally added space between the 2 classes  in my comment.
Also when using only one class the css is less specif and gets overridden.
This doesnt solve the issue for very small screens but I think it's ok (<950px)

MariaAga commented on webpack/assets/javascripts/react_app/components/Editor/editor.scss:
```diff
@@ -314,3 +314,14 @@
 }
+
+.autocompletion-checkbox {
+  margin-top: 0px;
+  margin-right: 20px;
+  margin-bottom: 5px;
+}
```
margin botton gets overwritten.
margin top I didn't see have any effect.
And margin right I think it looks better without it as it aligns with the dropboxes:
with `right:20px`
![Screenshot from 2021-09-07 14-55-41](https://user-images.githubusercontent.com/30431079/132348456-ca9ff1ed-2ad6-483c-9246-fbfb0017e5f9.png)
without
![Screenshot from 2021-09-07 14-55-33](https://user-images.githubusercontent.com/30431079/132348461-ff353f2a-d76e-479a-b099-b98ce2612610.png)

ofedoren replied:
I guess I'll remove those rules, I've tried to align the checkbox based on my preferences, but it works just fine with the default rules.

MariaAga reviewed (dismissed):
I have a weird thing happen, I get a console error `/snippets/ruby.js net::ERR_ABORTED 404 (Not Found)`, and it doesn't happen without this pr.
Also small css fix:

ofedoren commented:
Thanks, @MariaAga! Updated. I've removed `enableSnippets` since that error was somehow connected with this. I assume that if this flag was set then editor tries to load snippets for the language and since we don't want to use default ones (we don't use `import 'brace/snippets/ruby'`) it couldn't load them properly.

MariaAga commented:
Thanks @ofedoren, this looks good to me, @ezr-ondrej  Can you take a quick look over it?

ares commented:
I've retested the functionality and there's still one blocker for me. Given the autocompletion pops up even outside of the erb tags, it's very annoying to do general stuff. Try the following

1) edit Kickstart Default provisioning template
2) find the %packages section
3) try to add `httpd` and hit enter to add another package

the autocompletion will show you HostManaged#templates_used and when you hit the enter, it modified the `httpd` to this. This would be usability regression IMHO.

If we can't easily detect we're inside the ERB, can we instead allow the autocompletion to happen only on a special key combination? e.g. alt+enter would open the autocompletion pop up?

EDIT: I see we have ctrl + space to open it, so could it be the only time when the autocompletion is activated? I see the setting to enable/disable it entirely.

ofedoren commented:
Thanks, @ares :)

Well, I've tried to find a way to check if the cursor is within ERB tags, but I couldn't find any adequate solution, but to scan everything before the cursor until we find `%>` each time `getCompletions` is called. If anyone will have an idea, I'm open to it to implement it later :)

I've updated the PR, so the autocompletion will be shown only if `Ctrl + Space` is pressed by default. Users can change it via additional setting `Live Autocompletion`, which is `false` by default.

MariaAga commented on .eslintrc:
```diff
@@ -159,7 +159,9 @@
-          "cpu"
+          "cpu",
+          "Autocompletion",
+          "Live Autocompletion"
```
No need to add "Live Autocompletion" as live is a word and Autocompletion is defined above

ofedoren commented:
Thanks, @MariaAga :)

Updated.

ofedoren commented:
Tests are green, is there any other blocker, @ares? :)

adamruzicka commented:
Needs a rebase

ofedoren commented:
@adamruzicka, everything is :green_apple: again :)

ezr-ondrej reviewed (approved):
Awesome! IMHO the `Ctrl + Space` is the best way to go about it especially if user can disable it.
It is quite easy to miss tho.
Should we add some info about it into the `Help` tab? I'm fine doing it as follow-up. Also it is most certainly worth of a Headline feature.

ezr-ondrej commented:
@MariaAga were all your concerns addressed?

MariaAga commented:
@ezr-ondrej Yes :+1:

ezr-ondrej commented:
Lets get this in, great work @ofedoren ! Please demo this, it definitelly deserves to be shown :)

Thank you @ares, @MariaAga and @adamruzicka for reviews

ofedoren commented:
> Please demo this, it definitelly deserves to be shown :)
> 

I guess I've demoed this few times...

evgeni commented on webpack/assets/javascripts/react_app/components/Editor/EditorHelpers.js:
```diff
@@ -0,0 +1,159 @@
+} from 'lodash';
+
+import ace from 'brace';
+import 'brace/ext/language_tools';
```
```
[2022-01-11T09:32:31.759Z] ModuleNotFoundError: Module not found: Error: Can't resolve 'brace/ext/language_tools' in '/home/jenkins/workspace/foreman-packaging-deb-pr-test/debian/bullseye/build-foreman/foreman-3.2.0/webpack/assets/javascripts/react_app/components/Editor'
```

:crying_cat_face:

evgeni replied:
Same on RPM:
```
ModuleNotFoundError: Module not found: Error: Can't resolve 'brace/ext/language_tools' in '/builddir/build/BUILD/foreman-3.2.0-develop/webpack/assets/javascripts/react_app/components/Editor'
```

evgeni replied:
Seems `language_tools` is not exported from `vendor-core`?

https://github.com/theforeman/foreman-js/blob/1ee3ffa13a98fae2e839e3fb5c7cc69e00d1e503/packages/vendor-core/lib/modules.js#L75

ezr-ondrej replied:
Yeah, it seems so :thinking:
My bad for not checking, I'll see to fixing it :)

evgeni replied:
We really need a CI step that tries to compile stuff like prod does, without all those additional NPM packages.
